Implement exponential moving average estimator

awemany opened the pull request:
This is based on top of #1507 and extracts the moving average calculation into a new class to do it the right way(TM).

Also adds a simple unit test.

As per discussion with @dgenr8.

ptschip committed:
switch from a scoped_lock to a lock_guard in UpdateTransactionsPerSecond()

awemany committed:
Implement ExponentialMovingAverage stats class

To calculate time-wise exponential moving averages from discrete
events.

awemany committed:
Replace transaction rate calc in mempool with new EMA

sickpig commented:
Neat idea. Thanks @awemany and @dgenr8.

It seems like that we are failing while using clang to produce the binaries, see: https://travis-ci.org/BitcoinUnlimited/BitcoinUnlimited/jobs/463823492#L2267

```
../../src/test/stat_tests.cpp:151:38: error: too many arguments provided to
      function-like macro invocation
BOOST_AUTO_TEST_CASE(stat_expmovavg, *boost::unit_test::tolerance(1e-4))
                                     ^
/usr/include/boost/test/unit_test_suite.hpp:114:9: note: macro
      'BOOST_AUTO_TEST_CASE' defined here
#define BOOST_AUTO_TEST_CASE( test_name )                               \
        ^
../../src/test/stat_tests.cpp:151:1: error: C++ requires a type specifier for
      all declarations
BOOST_AUTO_TEST_CASE(stat_expmovavg, *boost::unit_test::tolerance(1e-4))
^
../../src/test/stat_tests.cpp:153:5: error: 'TestExponentialMovingAverage' does
      not refer to a value
    TestExponentialMovingAverage ema;
    ^
../../src/test/stat_tests.cpp:136:7: note: declared here
class TestExponentialMovingAverage : public ExponentialMovingAverage
      ^
../../src/test/stat_tests.cpp:176:2: error: expected ';' after top level
      declarator
}
 ^
 ;
```

awemany committed:
Make test of EMA compatible with boost 1.58

This commit can likely be reverted when upgrading to a more recent
version of libboost.

awemany commented:
@sickpig: Thanks for digging that out. It looks like this is due to me using too-new `libboost` features. I have added a commit that should make it compatible with `libboost1.58` again. Let's see how it goes.

awemany committed:
EMA: Fix a small chance of a race occuring

This is likely not an issue on 64bit arch but might be on 32bit.

awemany commented:
As a further test, I just confirmed its approximately correct behavior in the mempool TXN estimation, using

`./bitcoin-cli -regtest generate 1000`

and then `for i in `\``seq 0 100`\``; do ./bitcoin-cli -regtest sendtoaddress <bch-addr> 0.0001; ./bitcoin-cli -regtest getmempoolinfo; done`

building the txn rate up, followed by `watch ./bitcoin-cli -regtest getmempoolinfo` and watching the txn rate decay again.

dagurval commented on src/stat.cpp:
```diff
@@ -0,0 +1,32 @@
+
+ExponentialMovingAverage::ExponentialMovingAverage(const double _interval,
+    const size_t _time_units) :
+    interval(_interval), time_units(_time_units) {}
```
suggest throwing `std::invalid_argument` if `_interval == 0 || _time_units == 0`. You risk divide by zero in `::update` if not.

dagurval commented on src/stat.cpp:
```diff
@@ -0,0 +1,32 @@
+}
+
+void ExponentialMovingAverage::update(const size_t counts) {
+    std::lock_guard<std::mutex> lock(cs);
```
Nit: You could take the lock _after_ calling time `time()`

dagurval commented on src/test/stat_tests.cpp:
```diff
@@ -131,4 +132,47 @@ BOOST_AUTO_TEST_CASE(stat_empty_construct)
+    void set_time(const int64_t _t) { t = _t; }
+protected:
+    int64_t t = 0;
+    int64_t time() { return t; }
```
Nit: Add `override`

dagurval commented on src/stat.cpp:
```diff
@@ -0,0 +1,32 @@
+}
+
+double ExponentialMovingAverage::value() {
+    update(0); // decay rate to current value
```
A side effect in a getter function 🤔

dagurval commented on src/stat.h:
```diff
@@ -681,4 +681,47 @@ class LinearHistogram
+    //! Function used for time calculations. Can be overwritten e.g. for unit testing
+    virtual int64_t time();
+    //! Lock for serializing updates
+    std::mutex cs;
```
Nit: Make `cs` private
